grafana/resource_role_assignment.go

```diff
@@ -0,0 +1,186 @@
+	"context"
+
+	gapi "github.com/grafana/grafana-api-golang-client"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func ResourceRoleAssignment() *schema.Resource {
+	return &schema.Resource{
+		Description: `
+**Note:** This resource is available only with Grafana Enterprise 9.2+.
+* [Official documentation](https://grafana.com/docs/grafana/latest/enterprise/access-control/)
+* [HTTP API](https://grafana.com/docs/grafana/latest/http_api/access_control/)
+`,
+		CreateContext: UpdateRoleAssignments,
+		UpdateContext: UpdateRoleAssignments,
+		ReadContext:   ReadRoleAssignments,
+		DeleteContext: UpdateRoleAssignments,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"role_uid": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Grafana RBAC role UID.",
+			},
+			"users": {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Description: "Role assignments to users.",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:        schema.TypeInt,
+							Required:    true,
+							ForceNew:    false,
+							Description: "User ID.",
+						},
+						"global": {
+							Type:        schema.TypeBool,
+							Optional:    true,
+							Default:     false,
+							ForceNew:    false,
+							Description: "States whether the assignment is available across all organizations or not.",
+						},
+					},
+				},
+			},
+			"teams": {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Description: "Role assignments to teams.",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:        schema.TypeInt,
+							Required:    true,
+							ForceNew:    false,
+							Description: "Team ID.",
+						},
+					},
+				},
+			},
+			"service_accounts": {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Description: "Role assignments to service accounts.",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:        schema.TypeInt,
+							Required:    true,
+							ForceNew:    false,
+							Description: "Service account ID.",
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func ReadRoleAssignments(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client).gapi
+	uid := d.Get("role_uid").(string)
+	assignments, err := client.GetRoleAssignments(uid)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if err := setRoleAssignments(assignments, d); err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(uid)
+	return nil
+}
+
+func UpdateRoleAssignments(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	if !d.HasChange("users") && !d.HasChange("teams") && !d.HasChange("service_accounts") {
```

Comment:
hmm, you may need to check `d.IsNewResource()` as well for creation to work reliably in all cases
```suggestion
	if !d.IsNewResource() && !d.HasChange("users") && !d.HasChange("teams") && !d.HasChange("service_accounts") {
```